tests/test_send_recv_obj.py

```diff
@@ -172,3 +173,155 @@
+
+            frames = []
+
+            for i, (is_gpu, size) in enumerate(zip(is_gpus, sizes)):
+                if size > 0:
+                    resp = await self.ep.recv_obj(size, cuda=is_gpu)
+                else:
+                    resp = await self.ep.recv_future()
+                if is_gpu:
+                    breakpoint()
+                frame = ucp.get_obj_from_msg(resp)
+                frames.append(frame)
+
+            print(frames)
+            breakpoint()
+            res = cudf.Series(frames[-2]).to_pandas()
+            print(res)
+            return frames
+
+    class UCXListener:
+        def __init__(self):
+           self.comm_handler = None
+
+        def start(self):
+            async def serve_forever(ep, li):
+                print("starting server...")
+                ucx = UCX(ep)
+                self.comm = ucx
+
+            ucp.init()
+            loop = asyncio.get_event_loop()
+
+            self.ucp_server = ucp.start_listener(serve_forever,
+                                          listener_port=13337,
+                                          is_coroutine=True)
+            t = loop.create_task(self.ucp_server.coroutine)
+            self._t = t
+
+    uu = UCXListener()
+    uu.start()
+    uu.address = ucp.get_address()
+    uu.client = await ucp.get_endpoint(uu.address.encode(), 13337)
+    ucx = UCX(uu.client)
+    await ucx.write(cdf)
+    await asyncio.sleep(1)
+    frames = await uu.comm.read()
+
+    ucx_header = pickle.loads(frames[0])
+    ucx_index = bytes(frames[1])
+    cudf_buffer = frames[2]
+    ucx_bytes_leftovers = bytes(frames[3])
+    ucx_received_frames = [ucx_index, cudf_buffer, ucx_bytes_leftovers]
+    breakpoint()
+    res = cuda_deserialize(ucx_header, ucx_received_frames)
+    breakpoint()
+    res.to_pandas()# fails with NoneType
+    assert res.to_pandas() == cdf.to_pandas()
+    # assert cudf.Series(cudf_buffer).to_array() == cdf.to_array() failes with
+    # cudaMEmcpyDtoD
+
+    ucp.destroy_ep(uu.client)
+    ucp.fin()
+
+@pytest.mark.asyncio
+async def test_send_recv_cudf_small():
+    cudf = pytest.importorskip('cudf')
+    deserialize = pytest.importorskip("distributed.protocol").deserialize
+    serialize = pytest.importorskip("distributed.protocol").serialize
+    import struct
+    from distributed.utils import nbytes
+    import pickle
+
+    cuda_serialize = functools.partial(serialize, serializers=["cuda"])
+    cuda_deserialize = functools.partial(deserialize, deserializers=["cuda"])
+
+    cdf = cudf.Series([1, 2, 3])
+    header, frames = cuda_serialize(cdf)
+    cudf_frame = cdf.data.to_gpu_array()
+    size = len(cdf)
+
+    cuda_info = {'shape': [3,], 'typestr': '<i8'}
+
+    async with echo_pair(cuda_info) as (_, client):
+        await client.send_obj(bytes(str(size), encoding='utf-8'))
+        await client.send_obj(cudf_frame)
+        resp = await client.recv_obj(size, cuda=True)
+        result = ucp.get_obj_from_msg(resp)
+
```

Comment:
@Akshay-Venkatesh this test currently fails -- after converting the buffer back to a cupy/cudf object the array is [7, 0, 0] when it should be [1, 2, 3]

Reply:
Ben, this is a basic question but I'm not able to get cudf running. I tried the following:
```
python3 -m pip install cudf
```
But import fails with the following error:
```
akvenkatesh@prm-dgx-11:~/ucx-py$ python3
Python 3.7.2 (default, Mar  4 2019, 21:03:35) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cudf
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/akvenkatesh/py/install/lib/python3.7/site-packages/cudf/__init__.py", line 3, in <module>
    from cudf import dataframe
  File "/home/akvenkatesh/py/install/lib/python3.7/site-packages/cudf/dataframe/__init__.py", line 1, in <module>
    from cudf.dataframe import (buffer, dataframe, series,
  File "/home/akvenkatesh/py/install/lib/python3.7/site-packages/cudf/dataframe/dataframe.py", line 24, in <module>
    from cudf import formatting, _gdf
  File "/home/akvenkatesh/py/install/lib/python3.7/site-packages/cudf/_gdf.py", line 15, in <module>
    import nvcategory
  File "/home/akvenkatesh/py/install/lib/python3.7/site-packages/nvcategory.py", line 2, in <module>
    import pyniNVCategory
ImportError: libpython3.7m.so.1.0: cannot open shared object file: No such file or directory
```

Should I try building from source?

Reply:
Sorry I missed this -- we chatted briefly about installing from source and a small reproducible would be just using rmm:

```python
from librmm_cffi import librmm as rmm
import numpy as np
devicearray = rmm.to_device(np.array[1, 2, 3]))
```